{% extends "inchlib_base.html" %}

{% block content %}
{% load staticfiles %}
<script src="{% static "js/highcharts.js" %}"></script>
<!-- <script src="{% static "js/highcharts-more.js" %}"></script> -->

<script>
    $(document).ready(function(){

        window.boxplot = new Highcharts.Chart({

            chart: {
                renderTo: "boxplot",
                type: 'boxplot',
                height: 300
            },
            
            title: {
                text: null
            },
            
            legend: {
                enabled: false
            },
            exporting:{
                enabled: false,
            },
            xAxis: {
                categories: ["normal", "tumour"],
                title: {
                    text: 'Class'
                }
            },
            
            yAxis: {
                title: {
                    text: 'Average Signal'
                },
            },

            plotOptions: {
                boxplot: {
                    color: "#000000",
                    fillColor: '#F0F0E0',
                    lineWidth: 2,
                    medianColor: '#0C5DA5',
                    medianWidth: 3,
                    stemColor: '#A63400',
                    stemDashStyle: 'dot',
                    stemWidth: 1,
                    whiskerColor: '#3D9200',
                    whiskerLength: '20%',
                    whiskerWidth: 3 
                }
            },
        
            series: [{
                name: 'Average Signal',
                data: [
                ],
            }, 

            // {
            //     name: 'Outlier',
            //     color: Highcharts.getOptions().colors[0],
            //     type: 'scatter',
            //     data: [ // x, y positions where 0 is the first category
            //         [0, 644],
            //         [4, 718],
            //         [4, 951],
            //         [4, 969]
            //     ],
            //     marker: {
            //         fillColor: 'white',
            //         lineWidth: 1,
            //         lineColor: Highcharts.getOptions().colors[0]
            //     },
            //     tooltip: {
            //         pointFormat: 'Observation: {point.y}'
            //     }
            // }
            ]
            
        });

        var target_element = $("#microarrays_dendrogram");
        var offset = target_element.offset();
        var max_y = offset.top + 150;
        var x_pos = offset.left;
        var microarrays_element = $("#microarrays");
        microarrays_element.css({"left": x_pos + target_element.width()-50});

        window.dendrogram = new InCHlib(
            {
                target: "microarrays_dendrogram",
                dendrogram: true,
                max_height: 1500,
                font: "Trebuchet MS",
                heatmap_colors: "RdBkGr",
                column_dendrogram: true,
                header_as_heatmap_row: true,
                header_row_colors: "PiYG2",
                highlight_colors: "RdLrBu",
            });

        dendrogram.settings.row_onclick = function(row_ids, evt){
                    draw_boxplot(row_ids, evt);
                    dendrogram.highlight_rows(row_ids);
                };

        dendrogram.settings.dendrogram_node_onclick = function(row_ids, evt){
                    draw_boxplot(row_ids, evt);
                    dendrogram.highlight_rows([]);
                };

        dendrogram.read_data_from_file("{% static "data/microarrays.json" %}");
        dendrogram.draw();

        draw_boxplot(["ILMN_1737074"]);
        dendrogram.highlight_rows(["ILMN_1737074"]);

        function draw_boxplot(row_ids, evt){
            var class2values = {"tumour":[], "normal":[]};
            var i, j, values, row_id, done_rows = {};
            var header = dendrogram.data.header;

            for(i=0; i<row_ids.length; i++){
                row_id = dendrogram.objects2leaves[row_ids[i]];
                
                if(!(row_id in done_rows)){
                    values = dendrogram.data.nodes[row_id].features;
                    done_rows[row_id] = "";
                    for(j = 0; j<header.length; j++){
                        class2values[header[j]].push(values[j]);
                    }
                }
            }

            while(boxplot.series.length > 0){
                boxplot.series[0].remove(true);
            }

            var normal = get_boxplot_parameters(class2values.normal)
            var tumour = get_boxplot_parameters(class2values.tumour)

            boxplot.addSeries({data:[normal, tumour], name: "Average Signal"}, true);
        }

        function get_boxplot_parameters(array){
            var min, max, median, lower_q, upper_q;
            array.sort(function(a,b){return a - b;});
            var array_length = array.length;
            min = array[0];
            max = array[array_length-1];
            var median_index = dendrogram._hack_round(array_length/2);
            var quarter = dendrogram._hack_round(array_length/4);
            median = array[median_index];
            lower_q = array[median_index-quarter];
            upper_q = array[median_index+quarter];
            var parameters = [min, lower_q, median, upper_q, max]
            return parameters;
        }

});
</script>

<div class="pure-menu pure-menu-open pure-menu-horizontal">
    <ul>
        {% for e in examples %}
            {% if e.exampleid == example.exampleid %}
                <li class="pure-menu-selected"><a href="{% url use_cases e.exampleid %}">{{ e.title }}</a></li>
            {% else %}
                <li><a href="{% url use_cases e.exampleid %}">{{ e.title }}</a></li>
            {% endif %}
        {% endfor %}
    </ul>
</div>
<hr/>


<div class="description">
<p>DNA microarrays (DNA chip) is a tool which enables to simultaneously measure the expression levels of large numbers of genes. The differences in gene expressions (regulatory motifs) helps to discover the unique features common for various types of tested samples (e.g., malignant, benign, normal tissues).</p>

<p>The data set consist of 48 803 genes (compressed to 500) found on a microarray chip where expressions for 52 tumour/normal samples were measured.</p>

<p>The data were clustered using <i>Euclidean</i> distance and <i>complete</i> linkage. After the clustering the data were compressed from original 48 803 to 500 resulted data points. Rows found under the compression threshold were merged to a single row with a median value of original rows. </p>

<div class="small_title">Data set information</div>
<ul>
<li>instances: 500 genes (compressed from original 48 803)</li>
<li>features: 52 samples of carcinoma tissue (26) and normal mucosae (26)</li>
</ul>


<div class="small_title">Source: <a href='https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-1065/' target=blank>ArrayExpress - E-MTAB-1065 - Transcription profiling by array of clinical samples of carcinoma tissues compared to paired normal mucosae.</a></div>

<div class="citation"><a href="http://dx.doi.org/10.1111/boc.201200018" target="blank">Upregulation of IL-6, IL-8 and CXCL-1 production in dermal fibroblasts by normal/malignant epithelial cells in vitro: Immunohistochemical and transcriptomic analyses.</a> Kolář M, Szabo P, Dvořánková B, Lacina L, Gabius HJ, Strnad H, Sáchová J, Vlček C, Plzák J, Chovanec M, Cada Z, Betka J, Fík Z, Pačes J, Kovářová H, Motlík J, Jarkovská K, Smetana K Jr . Biology of the Cell 104(12):738-751 (2012), <a href="http://europepmc.org/abstract/MED/23043537" target="blank">Europe PMC 23043537</a></div>

<p>On click on the heatmap row or the dendrogram node the boxplot of values for individual classes (tumour/normal) is displayed next to the heatmap.</p>

</div>


<div id="microarrays">
        <div id="boxplot"></div>
</div>

<div id="microarrays_dendrogram"></div>
<hr/>

{% endblock %}